File: app/src/main/java/com/refreshme/data/Stylist.kt
package com.refreshme.data

import android.os.Parcelable
import com.google.firebase.firestore.Exclude
import com.google.firebase.firestore.GeoPoint
import com.google.firebase.firestore.IgnoreExtraProperties
import com.google.firebase.firestore.PropertyName
import kotlinx.parcelize.Parcelize
import kotlinx.parcelize.TypeParceler
import java.util.Date
import com.refreshme.data.GeoPointParceler 
import com.refreshme.data.DateParceler 

const val DEFAULT_AT_HOME_SERVICE_FEE = 20.0

@IgnoreExtraProperties
@Parcelize
data class FlashDeal(
    var title: String = "",
    var description: String = "",
    var discountPercentage: Int = 0,
    var expiryTime: Date? = null,
    var serviceId: String? = null
) : Parcelable

@IgnoreExtraProperties
@Parcelize
data class SocialMediaLinks(
    var instagram: String? = null,
    var tiktok: String? = null,
    var website: String? = null
) : Parcelable

@IgnoreExtraProperties
@Parcelize
@TypeParceler<GeoPoint?, GeoPointParceler>
@TypeParceler<Date?, DateParceler>
data class Stylist(
    var id: String = "",
    var name: String = "",

    @get:PropertyName("profileImageUrl") @set:PropertyName("profileImageUrl") var profileImageUrl: String? = "",
    @get:PropertyName("imageUrl") @set:PropertyName("imageUrl") var imageUrl: String? = "",

    var rating: Double = 0.0,
    @get:Exclude @set:Exclude var distance: Double = 0.0,
    
    @get:PropertyName("availableNow") @set:PropertyName("availableNow") var isAvailable: Boolean? = false,
    @get:PropertyName("available") @set:PropertyName("available") var availableInDb: Boolean? = false, 
    
    @get:PropertyName("location") @set:PropertyName("location") var location: GeoPoint? = null,
    
    @get:PropertyName("verified") @set:PropertyName("verified") var isVerified: Boolean? = false,
    @get:PropertyName("isVerified") @set:PropertyName("isVerified") var verifiedInDb: Boolean? = false, 
    
    var specialty: String? = "",
    var vibes: List<String>? = emptyList(), // Feature: Hyper-Local Vibe Search
    var recommendedFaceShapes: List<String>? = emptyList(), // Feature: AI Consultations
    var services: List<Service>? = emptyList(),
    var bio: String? = "",
    var portfolioImages: List<String>? = emptyList(),
    var portfolioVideos: List<String>? = emptyList(), // Feature: Video Portfolio Reels
    var beforeAfterImages: List<BeforeAfter>? = emptyList(),
    var reviews: List<Review>? = emptyList(),
    var address: String? = "",
    @get:PropertyName("salonAddress") @set:PropertyName("salonAddress") var salonAddress: String? = "",
    @get:PropertyName("serviceLocationType") @set:PropertyName("serviceLocationType") var serviceLocationType: String? = null,
    
    // Safety & Monetization (Mobile / 24-7)
    @get:PropertyName("atHomeServiceFee") @set:PropertyName("atHomeServiceFee") var atHomeServiceFee: Double? = 0.0,
    @get:PropertyName("emergencyFee") @set:PropertyName("emergencyFee") var emergencyFee: Double? = 0.0,
    @get:PropertyName("travelBufferMinutes") @set:PropertyName("travelBufferMinutes") var travelBufferMinutes: Int? = 30,
    @get:PropertyName("requiresIdVerificationForMobile") @set:PropertyName("requiresIdVerificationForMobile") var requiresIdVerificationForMobile: Boolean? = true,

    @get:PropertyName("offersAtHomeService") @set:PropertyName("offersAtHomeService") var offersAtHomeService: Boolean? = false,
    @get:PropertyName("maxTravelRangeKm") @set:PropertyName("maxTravelRangeKm") var maxTravelRangeKm: Int? = 15,
    @get:PropertyName("online") @set:PropertyName("online") var isOnline: Boolean? = false, 
    @get:PropertyName("offersEventBooking") @set:PropertyName("offersEventBooking") var offersEventBooking: Boolean? = false,

    var serviceType: ServiceType = ServiceType.IN_SALON,
    var verificationStatus: VerificationStatus = VerificationStatus.UNVERIFIED,
    var workingHours: List<WorkingHours>? = emptyList(),
    var bookedAppointments: List<Appointment>? = emptyList(),
    var reviewCount: Int? = 0,
    @get:PropertyName("featured") @set:PropertyName("featured") var isFeatured: Boolean? = false,
    var galleryImageUrls: List<String>? = emptyList(),
    @get:PropertyName("yearsOfExperience") @set:PropertyName("yearsOfExperience") var yearsOfExperience: Int? = 0,
    var tools: List<String>? = emptyList(),
    var stripeAccountId: String? = null,
    var stripeAccountStatus: String? = null,
    var subscriptionTier: String = "BASIC",
    @get:PropertyName("subscriptionActive") @set:PropertyName("subscriptionActive") var isSubscriptionActive: Boolean = false,
    @get:PropertyName("trialStartTime") @set:PropertyName("trialStartTime") var trialStartTime: Date? = null,
    @get:PropertyName("lastOnlineAt") @set:PropertyName("lastOnlineAt") var lastOnlineAt: Date? = null,
    var socialLinks: SocialMediaLinks? = null, // Feature: Profile Discovery & Enhancements
    var shopId: String? = null,
    var businessName: String? = null,
    var businessBio: String? = null,
    var businessAddress: String? = null,
    var businessPhone: String? = null,
    var businessWebsite: String? = null,
    var isShopProfile: Boolean = false,
    @get:Exclude @set:Exclude var matchScore: Int = 0, 
    @get:Exclude @set:Exclude var matchExplanation: String? = null,
    var currentFlashDeal: FlashDeal? = null, // Feature: Dynamic Flash Deals
    @get:PropertyName("servesGender") @set:PropertyName("servesGender")
    var servesGender: List<String>? = listOf("Men", "Women", "Non-binary"),

    /**
     * Professional categories this pro offers. Canonical ids come from
     * [StylistCategories] (hair / makeup / nails). Defaults to ["hair"] for
     * any pre-existing doc that was written before this field existed so
     * older stylists keep appearing on the Home list.
     */
    var categories: List<String>? = listOf(StylistCategories.HAIR)
) : Parcelable {
    
    @get:Exclude
    val displayImageUrl: String?
        get() = profileImageUrl?.ifBlank { imageUrl } ?: imageUrl

    @get:Exclude
    val hasPublicPhoto: Boolean
        get() = !displayImageUrl.isNullOrBlank() ||
            !portfolioImages.isNullOrEmpty() ||
            !portfolioVideos.isNullOrEmpty() ||
            !galleryImageUrls.isNullOrEmpty() ||
            beforeAfterImages.orEmpty().any {
                !it.beforeImageUrl.isNullOrBlank() || !it.afterImageUrl.isNullOrBlank()
            }

    @get:Exclude
    val hasPublicProfileSetup: Boolean
        get() = name.isNotBlank() &&
            (
                !specialty.isNullOrBlank() ||
                !bio.isNullOrBlank() ||
                !services.isNullOrEmpty() ||
                !displayAddress.isNullOrBlank() ||
                !businessName.isNullOrBlank() ||
                !businessBio.isNullOrBlank()
            )

    @get:Exclude
    val isDiscoverable: Boolean
        get() = isVerifiedStylist && hasPublicPhoto && hasPublicProfileSetup
    
    @get:Exclude
    val isVerifiedStylist: Boolean
        get() = isVerified == true || verifiedInDb == true
        
    @get:Exclude
    val isCurrentlyAvailable: Boolean
        get() = isAvailable == true || availableInDb == true

    @get:Exclude
    val offersMobileService: Boolean
        get() = offersAtHomeService == true ||
            serviceType == ServiceType.AT_HOME ||
            serviceType == ServiceType.ALL_HOURS ||
            serviceType == ServiceType.AFTER_HOURS

    @get:Exclude
    val displayAddress: String?
        get() = listOf(salonAddress, address, businessAddress)
            .firstOrNull { !it.isNullOrBlank() }

    @get:Exclude
    val hasFixedPublicLocation: Boolean
        get() {
            val loc = location ?: return false
            if (loc.latitude == 0.0 && loc.longitude == 0.0) return false
            val fixedByType = serviceLocationType == "fixed"
            val legacyFixedProfile = serviceLocationType.isNullOrBlank() && !offersMobileService
            return (fixedByType || legacyFixedProfile) && !displayAddress.isNullOrBlank()
        }

    @get:Exclude
    val effectiveAtHomeServiceFee: Double
        get() {
            val savedFee = atHomeServiceFee ?: 0.0
            return if (savedFee > 0.0) savedFee else if (offersMobileService) DEFAULT_AT_HOME_SERVICE_FEE else 0.0
        }

    @get:Exclude
    val hasActiveFlashDeal: Boolean
        get() {
            val deal = currentFlashDeal ?: return false
            val expiry = deal.expiryTime
            return expiry == null || expiry.after(Date())
        }
}

File: flutter_ios/lib/models/stylist.dart
/// RefreshMe — Stylist Model
/// Place this file at: lib/models/stylist.dart
///
/// Merge the stripeAccountId + stripeAccountStatus fields into your
/// existing Stylist model/class. Full example shown below.

import 'package:cloud_firestore/cloud_firestore.dart';

class Stylist {
  final String id;
  final String name;
  final String? profileImageUrl;
  final double rating;
  final bool isAvailable;
  final GeoPoint? location;
  final List<Service> services;
  final String? bio;
  final String? address;
  final String? salonAddress;
  final String? businessAddress;
  final String? serviceLocationType;
  final String? specialty;
  final bool offersAtHomeService;
  final double atHomeServiceFee;
  final int maxTravelRangeKm;

  // ── Stripe Connect fields (new) ──────────────────
  /// Stripe Connect Express account ID.
  /// Set automatically by Firebase webhook after onboarding.
  final String? stripeAccountId;

  /// "active" | "pending" | null
  final String? stripeAccountStatus;
  final bool? stripeChargesEnabled;
  final bool? stripePayoutsEnabled;
  final bool? stripeOnboardingComplete;
  // ─────────────────────────────────────────────────

  const Stylist({
    required this.id,
    required this.name,
    this.profileImageUrl,
    this.rating = 0.0,
    this.isAvailable = true,
    this.location,
    this.services = const [],
    this.bio,
    this.address,
    this.salonAddress,
    this.businessAddress,
    this.serviceLocationType,
    this.specialty,
    this.offersAtHomeService = false,
    this.atHomeServiceFee = 0.0,
    this.maxTravelRangeKm = 15,
    this.stripeAccountId,
    this.stripeAccountStatus,
    this.stripeChargesEnabled,
    this.stripePayoutsEnabled,
    this.stripeOnboardingComplete,
  });

  bool get hasActivePayoutAccount =>
      stripeAccountStatus == 'active' ||
      stripeOnboardingComplete == true ||
      (stripeChargesEnabled == true && stripePayoutsEnabled == true);

  String? get displayAddress {
    for (final value in [salonAddress, address, businessAddress]) {
      final trimmed = value?.trim();
      if (trimmed != null && trimmed.isNotEmpty) return trimmed;
    }
    return null;
  }

  bool get hasFixedPublicLocation {
    final loc = location;
    if (loc == null) return false;
    if (loc.latitude == 0.0 && loc.longitude == 0.0) return false;

    final fixedByType = serviceLocationType == 'fixed';
    final legacyFixedProfile =
        (serviceLocationType == null || serviceLocationType!.trim().isEmpty) &&
        !offersAtHomeService;
    return (fixedByType || legacyFixedProfile) &&
        (displayAddress?.isNotEmpty ?? false);
  }

  factory Stylist.fromFirestore(DocumentSnapshot doc) {
    final data = doc.data() as Map<String, dynamic>;
    return Stylist(
      id: doc.id,
      name: data['name'] as String? ?? '',
      profileImageUrl: data['profileImageUrl'] as String?,
      rating: (data['rating'] as num?)?.toDouble() ?? 0.0,
      isAvailable: data['isAvailable'] as bool? ?? true,
      location: data['location'] as GeoPoint?,
      services:
          (data['services'] as List<dynamic>?)
              ?.map((s) => Service.fromMap(s as Map<String, dynamic>))
              .toList() ??
          [],
      bio: data['bio'] as String?,
      address: data['address'] as String?,
      salonAddress: data['salonAddress'] as String?,
      businessAddress: data['businessAddress'] as String?,
      serviceLocationType: data['serviceLocationType'] as String?,
      specialty: data['specialty'] as String?,
      offersAtHomeService: data['offersAtHomeService'] as bool? ?? false,
      atHomeServiceFee: (data['atHomeServiceFee'] as num?)?.toDouble() ?? 0.0,
      maxTravelRangeKm: (data['maxTravelRangeKm'] as num?)?.toInt() ?? 15,
      stripeAccountId: data['stripeAccountId'] as String?,
      stripeAccountStatus: data['stripeAccountStatus'] as String?,
      stripeChargesEnabled: data['stripeChargesEnabled'] as bool?,
      stripePayoutsEnabled: data['stripePayoutsEnabled'] as bool?,
      stripeOnboardingComplete: data['stripeOnboardingComplete'] as bool?,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'name': name,
      'profileImageUrl': profileImageUrl,
      'rating': rating,
      'isAvailable': isAvailable,
      'location': location,
      'services': services.map((s) => s.toMap()).toList(),
      'bio': bio,
      'address': address,
      'salonAddress': salonAddress,
      'businessAddress': businessAddress,
      'serviceLocationType': serviceLocationType,
      'specialty': specialty,
      'offersAtHomeService': offersAtHomeService,
      'atHomeServiceFee': atHomeServiceFee,
      'maxTravelRangeKm': maxTravelRangeKm,
      'stripeAccountId': stripeAccountId,
      'stripeAccountStatus': stripeAccountStatus,
      'stripeChargesEnabled': stripeChargesEnabled,
      'stripePayoutsEnabled': stripePayoutsEnabled,
      'stripeOnboardingComplete': stripeOnboardingComplete,
    };
  }

  Stylist copyWith({
    String? id,
    String? name,
    String? profileImageUrl,
    double? rating,
    bool? isAvailable,
    GeoPoint? location,
    List<Service>? services,
    String? bio,
    String? address,
    String? salonAddress,
    String? businessAddress,
    String? serviceLocationType,
    String? specialty,
    bool? offersAtHomeService,
    double? atHomeServiceFee,
    int? maxTravelRangeKm,
    String? stripeAccountId,
    String? stripeAccountStatus,
    bool? stripeChargesEnabled,
    bool? stripePayoutsEnabled,
    bool? stripeOnboardingComplete,
  }) {
    return Stylist(
      id: id ?? this.id,
      name: name ?? this.name,
      profileImageUrl: profileImageUrl ?? this.profileImageUrl,
      rating: rating ?? this.rating,
      isAvailable: isAvailable ?? this.isAvailable,
      location: location ?? this.location,
      services: services ?? this.services,
      bio: bio ?? this.bio,
      address: address ?? this.address,
      salonAddress: salonAddress ?? this.salonAddress,
      businessAddress: businessAddress ?? this.businessAddress,
      serviceLocationType: serviceLocationType ?? this.serviceLocationType,
      specialty: specialty ?? this.specialty,
      offersAtHomeService: offersAtHomeService ?? this.offersAtHomeService,
      atHomeServiceFee: atHomeServiceFee ?? this.atHomeServiceFee,
      maxTravelRangeKm: maxTravelRangeKm ?? this.maxTravelRangeKm,
      stripeAccountId: stripeAccountId ?? this.stripeAccountId,
      stripeAccountStatus: stripeAccountStatus ?? this.stripeAccountStatus,
      stripeChargesEnabled: stripeChargesEnabled ?? this.stripeChargesEnabled,
      stripePayoutsEnabled: stripePayoutsEnabled ?? this.stripePayoutsEnabled,
      stripeOnboardingComplete:
          stripeOnboardingComplete ?? this.stripeOnboardingComplete,
    );
  }
}

class Service {
  final String id;
  final String name;
  final double price;
  final int durationMinutes;
  final String? description;
  final bool isBundle;
  final bool isAddOn;

  const Service({
    this.id = '',
    required this.name,
    required this.price,
    this.durationMinutes = 60,
    this.description,
    this.isBundle = false,
    this.isAddOn = false,
  });

  factory Service.fromMap(Map<String, dynamic> map) {
    return Service(
      id: map['id'] as String? ?? '',
      name: map['name'] as String? ?? '',
      price: (map['price'] as num?)?.toDouble() ?? 0.0,
      durationMinutes: map['durationMinutes'] as int? ?? 60,
      description: map['description'] as String?,
      isBundle: map['isBundle'] as bool? ?? map['bundle'] as bool? ?? false,
      isAddOn:
          map['isAddOn'] as bool? ??
          map['addOn'] as bool? ??
          _looksLikeAddOn(map['name'] as String? ?? ''),
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'name': name,
      'price': price,
      'durationMinutes': durationMinutes,
      'description': description,
      'isBundle': isBundle,
      'bundle': isBundle,
      'isAddOn': isAddOn,
      'addOn': isAddOn,
    };
  }

  static bool _looksLikeAddOn(String name) {
    final lower = name.toLowerCase();
    return lower.contains('add-on') || lower.contains('addon');
  }
}
